### ISSUE TITLE
[Discussion] Why did MMSU trigger pipeline bugs that SeedTTS didn't, despite identical I/O modalities?

### ISSUE THREAD

**Issue by zhaochenyang20** [sgl-project/sglang-omni#306]
## Context

Supporting the MMSU benchmark surfaced two non-trivial pipeline bugs back-to-back:

- **#285** — `EncoderInputPreparer.prepare()` concatenated Qwen3 audio tensors across requests with `torch.cat(..., dim=0)` without padding. Under concurrency, requests with different audio time lengths caused a tensor shape mismatch and crashed the encoder before reaching the audio tower.
- **#305** — `OmniEngine._filter_cached()` dropped cache-hit requests in mixed batches, leaving their futures unresolved and hanging the coordinator indefinitely. Trigger condition: the audio encoder batches a cache-hit request and a cache-miss request in the same scheduling step.

Before landing MMSU, my working assumption was that these paths were already well-exercised. From an **I/O modality perspective, MMSU and SeedTTS look equivalent** — audio/text in → text out. SeedTTS has been running this path for a while. I expected MMSU to be "swap the dataset and run."

It wasn't.

## The actually interesting question

Both fixes are merged / in review. But I think the question worth discussing is more useful than the fixes themselves:

> **Why didn't SeedTTS trigger these bugs, but MMSU did?**

If the two benchmarks are genuinely I/O-equivalent, one of the following must hold:

1. They are **not** actually equivalent — SeedTTS's request distribution happens to avoid these code paths along dimensions we weren't tracking (audio length variance, prompt redundancy, concurrency-induced batch composition, etc.)
2. Both benchmarks hit these paths, but SeedTTS's failure mode differs enough that we dismissed it as noise.

Looking at the two bugs concretely, I think (1) is clearly what happened, and the dimensions are quite specific:

**#285 — audio length variance.** The bug fires when concurrent requests carry audio features with different time dimensions. SeedTTS's TTS-style inputs tend to have more uniform length distributions per batch. MMSU audio clips (question stems + option prompts) vary much more in duration, so cross-request length heterogeneity becomes the common case rather than the exception.

**#305 — prompt redundancy × concurrency.** The bug fires when the encoder batches a cache-hit and a cache-miss in the same scheduling step. Whether this happens is essentially a function of **prompt redundancy in the dataset**:
- SeedTTS — each sample has a highly independent text/audio prompt → low cache hit rate → mixed batches are rare
- MMSU — multiple-choice questions share question stems and templated option structure → high cache hit rate → mixed batches become the common case

Notice the shape of both failures: they aren't modality bugs. They are **distribution bugs**. The code paths were always there; SeedTTS's request stream just never produced inputs that exercised them.

## Why this matters

If the analysis above is right, what we exposed isn't two isolated bugs. It's a **structural blind spot in our benchmark coverage strategy**: we organize benchmarks by modality combinations, but what actually determines code-path coverage is the **statistical properties of the request stream** — cache hit distribution, length variance, concurrency-induced batch composition. These two axes are orthogonal, and right now we only track one.

The current pattern is "new benchmark → round of firefighting." My concern is less about these specific bugs and more about the next workload (another benchmark, or real production traffic) hitting the same class of issue along yet another axis, and us going through the cycle again.

A few things I'd like input on:

1. Did the mixed-batch scenario in #305 or the length-mismatch scenario in #285 ever occur under SeedTTS? If not, was it because the relevant distributional property never showed up, or because they did occur and we dismissed the signal? Is there a cheap way to actually measure this rather than speculate?
2. Should we add a **stress-dimension** axis to how we classify benchmarks — not by dataset or modality, but by *cache hit rate*, *prompt diversity*, *audio/text length variance*, *batch composition under concurrency*? Before onboarding any new benchmark, we'd profile it along these axes to predict which code paths it will actually exercise.
3. Is there a low-cost way to synthesize these stress conditions in CI? For example, taking the SeedTTS dataset and (a) artificially inflating prompt repetition to force high cache hit rates, or (b) deliberately mixing short and long audio clips per batch — then running small concurrent tests. This would let us hit the distributional corners without needing to onboard a new benchmark first.

I'd rather invest one round of effort in getting the coverage framing right than keep paying this cost per benchmark.

cc @PopSoda2002 @Ratish1 — you both sit closest to these two root causes. Curious whether this framing matches your intuition, or whether you'd point at something else entirely.

**Comment by PasserBy4** [comment]
## Further investigation of #305

### 1. Why does the hanging-request issue appear when running the MMSU benchmark (text-only output, concurrency=8)?

Because there are cases where two samples share the same input audio. When the same audio appears again, the batch containing the second request at the audio encoder ends up with a mix of cached and uncached requests, which triggers the bug.

For example, when sampling 500 samples with `seed=42`, the following two samples share the same audio:

- `intonation_based_reasoning_cfeb892d-314e-421a-9cc3-60da7f69a39b`
- `intonation_perception_9deaed14-168e-4544-a7ab-77d77a843a9d`

When this audio appears for the first time, everything is normal:

<pre>
step=14 total=11 cached=0 uncached=11 mixed=False reqs=[56c3d29a:M:audio:529282ac6a <b>bc80dd1e:M:audio:68fc389486</b> 717b4825:M:audio:4cd1807245 173397d3:M:audio:68205139fe 03b9babf:M:audio:c717e6c1ee 5ef27e10:M:audio:7690411fb6 3bf30a29:M:audio:b3080c7b09 40420840:M:audio:f50727a0da 5f48fa1b:M:audio:708c7aad6d 87e57b8a:M:audio:9b5d08761b 55eb7505:M:audio:173ae0b139]
</pre>

(`M` indicates a cache miss, `H` indicates a cache hit. The bold entry is the first occurrence of the shared audio key, which gets written into the cache after this batch completes.)

When this audio appears for the second time, the cache hits, which turns the batch it is in into a mixed batch and triggers the bug:

<pre>
step=22 total=13 cached=1 uncached=12 mixed=True reqs=[47ff36c5:M:audio:5e1807f770 549c574e:M:audio:6bc6bac5a8 3592df2d:M:audio:4f3eae1cd5 ea60fc5a:M:audio:6295dc8683 6086f9ed:M:audio:be79b36623 <b>22fc2b42:H:audio:68fc389486</b> 46ffc0b3:M:audio:cd4b15c03b 2065bc96:M:audio:3f5962744c 61a3b207:M:audio:34bbe8dddb d65693e8:M:audio:24a507717b 7e212272:M:audio:1bcea1ca3b e3112868:M:audio:49f46a827a 791dd393:M:audio:71f11207fd]
</pre>

The fact that the same dataset occasionally does not reproduce the bug is because the second occurrence of the repeated audio happens to be alone in its batch, avoiding mixing with other uncached requests.

### 2. Why was this issue not observed when running SeedTTS voice clone (text+audio output, concurrency=8)?

In fact, SeedTTS has a large amount of shared-audio reuse, whereas MMSU only has a small amount. The reason this was not observed before is that testing SeedTTS voice clone always has talker output enabled. Once audio output is on, the downstream `code_predictor` and `code2wav` are serial, which forces requests to complete one by one. As a result the semaphore is released one slot at a time, upstream requests arrive at the audio encoder one by one, and they cannot form a batch (the system is in a pseudo-parallel, effectively serial state). Since `batch_size = 1` at the audio encoder on every call, a mix of cached and uncached requests essentially never occurs:

<pre>
step=1 total=1 cached=0 uncached=1 mixed=False reqs=[<b>b313a491:M:audio:be8aa5038d</b>]
step=2 total=7 cached=1 uncached=6 mixed=True reqs=[f526a496:M:audio:5f2ee3bf07 <b>54689342:H:audio:be8aa5038d</b> 3f9acabb:M:audio:6108bd95af fa7174f3:M:audio:146621ede0 f2cd1c0f:M:audio:6108bd95af 1dd9830c:M:audio:146621ede0 32c458a1:M:audio:5f2ee3bf07]
step=28387 total=1 cached=0 uncached=1 mixed=False reqs=[<b>cfa4bed5:M:audio:39ba9bd50f</b>]
step=33037 total=1 cached=1 uncached=0 mixed=False reqs=[<b>7be8debc:H:audio:39ba9bd50f</b>]
step=37095 total=1 cached=0 uncached=1 mixed=False reqs=[<b>a4e7a54a:M:audio:879f496124</b>]
step=38641 total=1 cached=1 uncached=0 mixed=False reqs=[<b>2c721dd3:H:audio:879f496124</b>]
step=42458 total=1 cached=0 uncached=1 mixed=False reqs=[cc517767:M:audio:dae294e842]
</pre>

As can be seen from the example above, in step 1 one request entered the audio encoder first because its preprocess finished faster; then the remaining 7 requests entered the audio encoder as a single batch. One of those requests triggered a cache hit, which caused `mixed=True` at step 2 and produced one occurrence of the bug. However, this only happens to the first wave of 8 requests. All subsequent requests come in one by one, cannot form a batch, and the same situation never appears again.

Therefore, if we enable audio output for MMSU, the same situation will occur (because speech generation is serial, the semaphore is released one slot at a time, new requests arrive one by one, and the bug is avoided):

```
06:21:04.964 step=1 total=1 cached=0 uncached=1 mixed=False reqs=[5fccdb24:M:audio:fa1a0ee3d2]
06:21:05.467 step=2 total=7 cached=0 uncached=7 mixed=False reqs=[ea7d0047:M:audio:ec566f5487 3da6045c:M:audio:ba8a51044a 26e755df:M:audio:cd9a0f783d 444a9990:M:audio:5282dc2c6d bed96b37:M:audio:d32be5ae28 280795a8:M:audio:c5a6b2ec2e 05321ee8:M:audio:b2e07ec2a5]
06:21:20.113 step=3 total=1 cached=0 uncached=1 mixed=False reqs=[596f791b:M:audio:3e1f80cd09]
06:21:21.746 step=4 total=1 cached=0 uncached=1 mixed=False reqs=[848fad37:M:audio:44da56ec9a]
06:21:21.867 step=5 total=1 cached=0 uncached=1 mixed=False reqs=[50c7a53f:M:audio:6c5f9b1a2f]
```

### Summary

The trigger of this bug is that multiple samples share the same audio, which causes a mix of cached and uncached requests within a single audio encoder batch. The reason the issue was not observed previously when testing voice clone is that audio output was enabled, which forces the system to run serially; the audio encoder's `batch_size` is always 1, so `mixed=True` cannot be triggered.

cc @zhaochenyang20 @PopSoda2002

**Comment by zhaochenyang20** [comment]
https://github.com/sgl-project/sglang-omni/pull/305#issuecomment-4269853189

Good!

**Comment by Ratish1** [comment]
First, thanks to @PopSoda2002 for the root-cause analysis in #305: https://github.com/sgl-project/sglang-omni/pull/305#issuecomment-4269853189

Also thanks to @PasserBy4  for the follow-up investigation in this issue.

Hey @zhaochenyang20 , after rechecking the code path and rerunning the benchmark comparison, my current understanding is:

- The `#305` explanation is correct: the hang is triggered by a mixed cache-hit/cache-miss batch, not by MMSU specifically.
- In the MMSU 500-sample `seed=42` run, there is exactly one duplicated audio pair, and the hanging sample is the second occurrence of that pair.
- That confirms the repeated-audio case is enough to create the mixed cache-hit/cache-miss condition under concurrency.

What the comparison also shows is that the broader MMSU vs SeedTTS difference is not explained by shared-audio reuse alone:

- MMSU 500-sample `seed=42` subset:
  - 499 unique audios
  - 1 duplicate group
  - 2 duplicate samples total
- SeedTTS-50:
  - 32 unique audios
  - 18 duplicate groups
  - 36 duplicate samples total

So in this comparison, SeedTTS actually has much more reuse.

The stronger difference is cross-request audio-length variance plus runtime topology:

- MMSU has much larger cross-request audio duration spread:
  - max window spread of 25.50s
- SeedTTS-50 is much tighter:
  - max window spread of 3.98s

And on the runtime side:

- thinker-only MMSU text reproduces the hang
- speech-enabled MMSU text+audio completes cleanly
- SeedTTS plain / VC at concurrency 4 also complete cleanly

So my current read is:

- `#305` is a mixed cache-hit/cache-miss scheduling bug, and the repeated MMSU audio pair is enough to trigger it.
- `#285` is a cross-request variable-length audio batching bug, and MMSU is much more likely to expose that because of the larger audio-length variance.
- The broader MMSU vs SeedTTS difference seems better explained by:
  - sparse repeated-audio trigger for `#305`
  - much higher cross-request audio-length variance for `#285`
  - and different runtime topology between thinker-only MMSU and speech-enabled SeedTTS / MMSU

So this looks more like different workloads exposing different stress dimensions, rather than an MMSU-specific vs SeedTTS-specific behavior difference.

**Comment by PopSoda2002** [comment]
I think this issue can be closed 💯

**Comment by zhaochenyang20** [comment]
great work my man! @PopSoda2002 @PasserBy4

**Comment by zhaochenyang20** [comment]
Have a linkedin post:

https://www.linkedin.com/feed/update/urn:li:activity:7451116096751026176/

@PasserBy4 @PopSoda2002 @Ratish1 thanks so much guys
